catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/builder/AttributeValueNormalizer.java

```diff
@@ -0,0 +1,139 @@
+ */
+package org.codice.ddf.catalog.ui.forms.builder;
+
+import static ddf.catalog.data.AttributeType.AttributeFormat.DATE;
+
+import ddf.catalog.data.AttributeDescriptor;
+import ddf.catalog.data.AttributeRegistry;
+import java.time.DateTimeException;
+import java.time.Instant;
+import java.time.OffsetDateTime;
+import java.time.format.DateTimeParseException;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import javax.annotation.Nullable;
+import org.codice.ddf.catalog.ui.forms.filter.FilterProcessingException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Normalization for values on Filter JSON.
+ *
+ * <p>Input {@link #normalizeForXml(String, String)} should allow exceptions to propagate. Output
+ * {@link #normalizeForJson(String, String)} should be more forgiving so the feature is hardened
+ * against failure, since the output function reads directly from the database.
+ *
+ * <p>Property validation is done <b>only</b> on data entering the system, not exiting.
+ */
+class AttributeValueNormalizer {
+  private static final Logger LOGGER = LoggerFactory.getLogger(AttributeValueNormalizer.class);
+
+  private static final Pattern EXPECTED_ATTRIBUTE_NAME_PATTERN =
+      Pattern.compile("(ext\\.)?(\\p{Alpha}+\\.)?\\p{Alpha}+(-\\p{Alpha}+)*");
+
+  private final AttributeRegistry registry;
+
+  AttributeValueNormalizer(AttributeRegistry registry) {
+    this.registry = registry;
+  }
+
+  String normalizeForJson(String property, String value) {
+    if (eitherStringIsNull(property, value)) {
+      return value;
+    }
+    if (isNotNormalizableDateValue(property, value)) {
+      return value;
+    }
+    Instant instant = instantFromEpoch(value);
+    if (instant != null) {
+      return instant.toString();
+    }
+    return value;
+  }
+
+  String normalizeForXml(String property, String value) {
+    if (eitherStringIsNull(property, value)) {
+      return value;
+    }
+    if (!EXPECTED_ATTRIBUTE_NAME_PATTERN.matcher(property).matches()) {
+      throw new FilterProcessingException("Malformed attribute name on search form: " + property);
+    }
+    if (isNotNormalizableDateValue(property, value)) {
+      return value;
+    }
+    Instant epoch = instantFromEpoch(value);
+    if (epoch != null) {
+      return value;
+    }
+    Instant iso = instantFromIso(value);
+    if (iso != null) {
+      return Objects.toString(iso.toEpochMilli());
+    }
+    throw new FilterProcessingException("Unexpected date format on search form: " + value);
+  }
+
+  private boolean eitherStringIsNull(String property, String value) {
+    if (property == null || value == null) {
+      LOGGER.trace("Property [{}] or value [{}] was null, ignoring normalization", property, value);
+      return true;
+    }
+    return false;
+  }
+
+  private boolean isNotNormalizableDateValue(String property, String value) {
+    Optional<AttributeDescriptor> optional = registry.lookup(property);
```

Comment:
✏️ `.map` and other `Optional` methods should be preferred to `isPresent`.

Reply:
Notice I want to return early if the optional is empty. I'm not sure there's another way to accomplish that. I'm checking for `!optional.isPresent()`.

Reply:
You could do:
```
optional
  .map(descriptor -> descriptor.getType().getAttributeFormat() != DATE)
  .orElse(true);
```